dashboard/memory_utils.py

```diff
@@ -0,0 +1,302 @@
+        if self.pinned_in_memory:
+            return ReferenceType.PINNED_IN_MEMORY
+        elif self.submitted_task_ref_count > 0:
+            return ReferenceType.USED_BY_PENDING_TASK
+        elif self.local_ref_count > 0:
+            return ReferenceType.LOCAL_REFERENCE
+        elif len(self.contained_in_owned) > 0:
+            return ReferenceType.CAPTURED_IN_OBJECT
+        else:
+            return ReferenceType.UNKNOWN_STATUS
+
+    def _is_object_ref_actor_handle(self) -> bool:
+        object_ref_hex = self.object_ref.hex()
+
+        # random (8B) | ActorID(6B) | flag (2B) | index (6B)
+        # ActorID(6B) == ActorRandomByte(4B) + JobID(2B)
+        # If random bytes are all 'f', but ActorRandomBytes
+        # are not all 'f', that means it is an actor creation
+        # task, which is an actor handle.
+        random_bits = object_ref_hex[:TASKID_RANDOM_BITS_SIZE]
+        actor_random_bits = object_ref_hex[TASKID_RANDOM_BITS_SIZE:
+                                           TASKID_RANDOM_BITS_SIZE +
+                                           ACTORID_RANDOM_BITS_SIZE]
+        if (random_bits == "f" * 16 and not actor_random_bits == "f" * 8):
+            return True
+        else:
+            return False
+
+    def __dict__(self):
+        return {
+            "object_ref": self.object_ref.hex(),
+            "pid": self.pid,
+            "node_ip_address": self.node_address,
+            "object_size": self.object_size,
+            "reference_type": self.reference_type,
+            "call_site": self.call_site,
+            "local_ref_count": self.local_ref_count,
+            "pinned_in_memory": self.pinned_in_memory,
+            "submitted_task_ref_count": self.submitted_task_ref_count,
+            "contained_in_owned": [
+                object_ref.hex() for object_ref in self.contained_in_owned
+            ],
+            "type": "Driver" if self.is_driver else "Worker"
+        }
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        return str(self.__dict__())
+
+
+class MemoryTable:
+    def __init__(self,
+                 entries: List[MemoryTableEntry],
+                 group_by_type: GroupByType = GroupByType.NODE_ADDRESS,
+                 sort_by_type: SortingType = SortingType.PID):
+        self.table = entries
+        # Group is a list of memory tables grouped by a group key.
+        self.group = {}
+        self.summary = defaultdict(int)
+        if group_by_type and sort_by_type:
+            self.setup(group_by_type, sort_by_type)
+        elif group_by_type:
+            self._group_by(group_by_type)
+        elif sort_by_type:
+            self._sort_by(sort_by_type)
+
+    def setup(self, group_by_type: GroupByType, sort_by_type: SortingType):
+        """Setup memory table.
+
+        This will sort entries first and gruop them after.
+        Sort order will be still kept.
+        """
+        self._sort_by(sort_by_type)._group_by(group_by_type)
```

Comment:
Is the `self._sort_by(...)._group_by(...)` equals with `self._group_by(...)._sort_by(...)`?

Reply:
It is not equal, which I agree is somewhat confusing. That said, it is sorted before grouped in all code paths.